src/server.c

```diff
@@ -680,6 +680,52 @@ int clientsCronResizeQueryBuffer(client *c) {
     return 0;
 }
 
+/* The client output buffer can be adjusted to better fit the memory requierments.
+ *
+ * the logic is:
+ * in case the last observed peak size of the buffer equals the buffer size - we double the size
+ * in case the last observed peak size of the buffer is less than half the buffer size - we shrink by half.
+ * The buffer peak will be reset back to zero every 5 seconds
+ * The function always returns 0 as it never terminates the client. */
+int clientsCronResizeOutputBuffer(client *c) {
+
+    size_t new_buffer_size = 0;
+    const size_t buffer_target_shrink_size = c->buf_usable_size/2;
+    const size_t buffer_target_expend_size = c->buf_usable_size*2;
+
+    if (buffer_target_shrink_size >= PROTO_REPLY_MIN_BYTES &&
+        c->buf_peak < buffer_target_shrink_size ) {
+        new_buffer_size = buffer_target_shrink_size;
+        server.stat_reply_buffer_shrinks++;
+    } else if(buffer_target_expend_size <= PROTO_REPLY_CHUNK_BYTES &&
+            c->buf_peak == c->buf_usable_size) {
+        new_buffer_size = buffer_target_expend_size;
+        server.stat_reply_buffer_expends++;
+    }
+
+    /* reset the peak value each 5 seconds. in case the client will be idle
+     * it will start to shrink.
+     */
+    if(server.unixtime - c->buf_peak_last_reset_time >= 5) {
+        c->buf_peak = 0UL;
+        c->buf_peak_last_reset_time = server.unixtime;
+    }
+
+    if(new_buffer_size) {
+        /* In case the buffer possition in 0 we can just delete and realloc the buffer with the new size
+         * no memmove is needed.
+         */
+        if(c->bufpos == 0) {
+            zfree(c->buf);
+            c->buf = zmalloc(new_buffer_size);
+            c->buf_usable_size = zmalloc_usable_size(c->buf);
```

Comment:
since you saved one call by using zrealloc_usable, we can do that here too.
```suggestion
            c->buf = zmalloc_usable(new_buffer_size, c->buf_usable_size);
```